### ISSUE TITLE
Make update is broken

### ISSUE THREAD

**Issue by beldmit** [openssl/openssl#12891]
Applying patch (desired to fix #12735)
[example.txt](https://github.com/openssl/openssl/files/5230758/example.txt)
```
patch -p1 < example.txt
make update
make clean
make
```

get the following error
```
In file included from ssl/ssl_err.c:11:
ssl/ssl_err.c:315:31: error: 'SSL_R_OCSP_CALLBACK_FAILURE' undeclared here (not in a function); did you mean 'SSL_R_CALLBACK_FAILED'?
     {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_OCSP_CALLBACK_FAILURE),
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~
include/openssl/err.h:256:25: note: in definition of macro 'ERR_PACK'
       (((unsigned long)(reason) & ERR_REASON_MASK)) )
                         ^~~~~~
ssl/ssl_err.c:317:31: error: 'SSL_R_OCSP_RESPONSE_IS_INVALID' undeclared here (not in a function); did you mean 'RSA_R_LAST_OCTET_INVALID'?
     {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_OCSP_RESPONSE_IS_INVALID),
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
include/openssl/err.h:256:25: note: in definition of macro 'ERR_PACK'
       (((unsigned long)(reason) & ERR_REASON_MASK)) )
                         ^~~~~~
make[1]: *** [Makefile:18658: ssl/libssl-lib-ssl_err.o] Error 1
make[1]: Leaving directory '/home/beldmit/upstream/openssl'
make: *** [Makefile:3126: build_sw] Error 2
```

The resulting diff of `util/libssl.num` after `make update` invocation seems unreasonably huge marking a lot of function as non-existent.

**Comment by levitte** [comment]
Was it *only* that patch?  A couple of new `SSL_R_` macros shouldn't affect util/libssl.num, at all.

**Comment by levitte** [comment]
Note that I found a but in OpenSSL::ParseC, fixed in #12882

**Comment by beldmit** [comment]
I just merged the current master branch to my copy without any other changes.

Should I try the fix you refer to?

**Comment by beldmit** [comment]
Tried and got the same error :(

**Comment by levitte** [comment]
Can you point to an exact starting point?  Repository and commit id?

**Comment by beldmit** [comment]
Repo:
https://github.com/beldmit/openssl

Head:
commit 4343a4187d28d719006642a610afea6e186832bf (HEAD -> master, upstream/master, origin/master, origin/HEAD)
Author: Shane Lontis <shane.lontis@oracle.com>
Date:   Wed Sep 16 14:10:23 2020 +1000

**Comment by levitte** [comment]
So, regarding the compilation problem you get, this is the diff that `make update` created:

``` C
diff --git a/ssl/ssl_err.c b/ssl/ssl_err.c
index 9f47a924f0..ba245bfe09 100644
--- a/ssl/ssl_err.c
+++ b/ssl/ssl_err.c
@@ -9,7 +9,7 @@
  */
 
 #include <openssl/err.h>
-#include <openssl/sslerr.h>
+#include <openssl/ssl.h>
 
 #ifndef OPENSSL_NO_ERR
 
@@ -312,6 +312,10 @@ static const ERR_STRING_DATA SSL_str_reasons[] = {
     {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_NULL_SSL_CTX), "null ssl ctx"},
     {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_NULL_SSL_METHOD_PASSED),
     "null ssl method passed"},
+    {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_OCSP_CALLBACK_FAILURE),
+    "ocsp callback failure"},
+    {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_OCSP_RESPONSE_IS_INVALID),
+    "ocsp response is invalid"},
     {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_OLD_SESSION_CIPHER_NOT_RETURNED),
     "old session cipher not returned"},
     {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_OLD_SESSION_COMPRESSION_ALGORITHM_NOT_RETURNED),
```

See the changed `#include`?  There's your problem, and that's `mkerr.pl` making certain assumptions that work for all error stuff...  except the SSL one.  A simple solution would be to have `ssl.h` include `sslerr.h`, which is the pattern we have elsewhere.

The reason we haven't seen this for a while?  I guess we haven't added a new `SSL_R_` error in a while...

As to the libssl.num problem, I'm still trying to figure that one out.  I assume it's some sort of parsing problem...  it seems to be triggered only when you build in the source tree (I habitually have a separate build tree, so...)

**Comment by beldmit** [comment]
I see the include change. AFAIR, I tried to revert it but still got the same error.

**Comment by beldmit** [comment]
I reverted the #include replacement.
The situation did not become better.

```
In file included from ssl/ssl_err.c:11:
ssl/ssl_err.c:315:31: error: 'SSL_R_OCSP_CALLBACK_FAILURE' undeclared here (not in a function); did you mean 'SSL_R_CALLBACK_FAILED'?
     {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_OCSP_CALLBACK_FAILURE),
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~
include/openssl/err.h:256:25: note: in definition of macro 'ERR_PACK'
       (((unsigned long)(reason) & ERR_REASON_MASK)) )
                         ^~~~~~
ssl/ssl_err.c:317:31: error: 'SSL_R_OCSP_RESPONSE_IS_INVALID' undeclared here (not in a function); did you mean 'SSL_R_COMPRESSION_DISABLED'?
     {ERR_PACK(ERR_LIB_SSL, 0, SSL_R_OCSP_RESPONSE_IS_INVALID),
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
include/openssl/err.h:256:25: note: in definition of macro 'ERR_PACK'
       (((unsigned long)(reason) & ERR_REASON_MASK)) )
                         ^~~~~~
make[1]: *** [Makefile:18651: ssl/libssl-lib-ssl_err.o] Error 1
make[1]: Leaving directory '/home/beldmit/upstream/openssl'
make: *** [Makefile:3119: build_sw] Error 2
```

**Comment by levitte** [comment]
Huh, it seems it doesn't touch include/openssl/sslerr.h...  wonder why

**Comment by levitte** [comment]
I'll get back on this...  for the moment, I need my mind elsewhere

**Comment by levitte** [comment]
I found that something seems very wrong...  `util/mkerr.pl` will suddenly write to `include/openssl/foo.h` instead of `include/openssl/fooerr.h`, and for the moment, I can't even see where that file rename before writing is supposed to happen

**Comment by levitte** [comment]
Fix in #12910.
